# 4. EDA with Visualization

Charts exploring how FlightNumber, PayloadMass, LaunchSite, and Orbit relate
to first-stage landing success.

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

df = pd.read_csv('../data/spacex_launch_data.csv')
sns.set_style('whitegrid')

## 4.1 Flight number vs. payload mass, colored by landing outcome

In [ ]:
plt.figure(figsize=(10, 6))
sns.scatterplot(data=df, x='FlightNumber', y='PayloadMass', hue='Class', palette={0: 'crimson', 1: 'seagreen'})
plt.xlabel('Flight Number')
plt.ylabel('Payload Mass (kg)')
plt.title('Flight Number vs. Payload Mass, colored by landing outcome')
plt.show()

**Observation:** Landing success (green) becomes denser at higher flight numbers —
later missions land more reliably. Very heavy payloads (>10,000 kg) skew toward
failed landings (red), consistent with GTO missions needing more of the booster's
fuel for the ascent, leaving less margin for the landing burn.

## 4.2 Success rate by launch site

In [ ]:
site_rate = df.groupby('LaunchSite')['Class'].mean().sort_values(ascending=False)
print(site_rate)

plt.figure(figsize=(8, 5))
sns.barplot(x=site_rate.index, y=site_rate.values, palette='viridis')
plt.ylabel('Success rate')
plt.title('Landing success rate by launch site')
plt.ylim(0, 1)
plt.show()

LaunchSite
KSC LC 39A      0.772727
VAFB SLC 4E     0.769231
CCAFS SLC 40    0.600000
Name: Class, dtype: float64


## 4.3 Success rate by orbit type

In [ ]:
orbit_rate = df.groupby('Orbit')['Class'].mean().sort_values(ascending=False)
print(orbit_rate)

plt.figure(figsize=(10, 5))
sns.barplot(x=orbit_rate.index, y=orbit_rate.values, palette='mako')
plt.ylabel('Success rate')
plt.title('Landing success rate by orbit type')
plt.xticks(rotation=45)
plt.show()

Orbit
ES-L1    1.000000
GEO      1.000000
HEO      1.000000
SSO      1.000000
VLEO     0.857143
LEO      0.714286
PO       0.666667
MEO      0.666667
ISS      0.619048
GTO      0.518519
SO       0.000000
Name: Class, dtype: float64


## 4.4 Yearly trend in success rate

In [ ]:
df['Year'] = pd.to_datetime(df['Date']).dt.year
yearly_rate = df.groupby('Year')['Class'].mean()
print(yearly_rate)

plt.figure(figsize=(10, 5))
plt.plot(yearly_rate.index, yearly_rate.values, marker='o', color='darkorange', linewidth=2)
plt.xlabel('Year')
plt.ylabel('Success rate')
plt.title('Landing success rate by year')
plt.ylim(0, 1.05)
plt.grid(alpha=0.3)
plt.show()

**Summary of EDA findings:**
- Success rate climbs from **58%** in the first half of the launch history to **76%** in the second half.
- **KSC LC-39A** and **VAFB SLC-4E** substantially outperform **CCAFS SLC-40** (~77% vs. 60%).
- Orbits requiring the most delta-v (**GTO**) have the lowest success rate (52%); low-energy orbits (**VLEO, SSO, GEO**) land far more reliably.
- These patterns motivate including `FlightNumber`, `LaunchSite`, `Orbit`, and `PayloadMass` as key features in the classification model (notebook 6).